### ISSUE TITLE
[SETUP-BUG] Problem setting up Scenario 5 using Minio server to store artifacts

### ISSUE THREAD

**Issue by rgaudenzi-fub** [mlflow/mlflow#5514]
Good evening, I'm trying to setup a MLflow deploy based on docker container following the structure depicted in Scenario 5 (https://mlflow.org/docs/latest/tracking.html#scenario-5-mlflow-tracking-server-enabled-with-proxied-artifact-storage-access) where MLflow acts as artifact proxy. 
My idea was to have three container:
- a MySQL container to store data about runs, metrics and parameters;
- a container running Minio where MLflow will store my artifacts;
- a container running MLflow.

I don't want to expose the minio over Internet, but I want to make it accessible only from the MLflow server, and I thought that I could do that after seeing the Scenario 5 description in the documentation.

In the MLflow container I set up the envinroment variables:
MLFLOW_S3_ENDPOINT_URL, 
AWS_ACCESS_KEY_ID 
AWS_SECRET_ACCESS_KEY 

then I created a bucket inside Minio named "mlflow", and then start the tracking server in this way:
mlflow server \
--backend-store-uri mysql+mysqlconnector://$DB_USER:$DB_PWD@db:3306/mlflowtest \
--artifacts-destination s3://mlflow \
--serve-artifacts \
--host 0.0.0.0 \
--gunicorn-opts "--log-level debug"

At this point, I have a python script that does the following:
mlflow.set_tracking_uri('http://localhost:5000/')
mlflow.create_experiment("test")
mlflow.set_experiment("test")

with mlflow.start_run():
    print("Test LOG ARTIFACT - start")
    mlflow.log_artifact(local_path="./data/language_detection.csv")
    print("Test LOG ARTIFACT - done")

But I get the following error:
mlflow.exceptions.MlflowException: Could not find a registered artifact repository for: mlflow-artifacts:/39/668c9fe202184565897f08f6234049be/artifacts. Currently registered schemes are: ['', 'file', 's3', 'gs', 'wasbs', 'ftp', 'sftp', 'dbfs', 'hdfs', 'viewfs', 'runs', 'models']

I've also tried to setup the parameter "default-artifact-root" in some ways:
- If I set it to "--default-artifact-root s3://mlflow" then I get the error "Unable to locate credentials". If I set the envinroment variables inside the python client code, then I get the error "boto3.exceptions.S3UploadFailedError: Failed to upload ./data/language_detection.csv to mlflow/a8df2597bad64420af0ef56a19708bee/artifacts/language_detection.csv: An error occurred (InvalidAccessKeyId) when calling the PutObject operation: The AWS Access Key Id you provided does not exist in our records."
- If I set it to "http://localhost:5000/api/2.0/mlflow-artifacts/artifacts/myruns", then I get the error: mlflow.exceptions.MlflowException: Could not find a registered artifact repository for: http://localhost:5000/api/2.0/mlflow-artifacts/artifacts/myruns/37/7e27c1a84966472f9c871e9a62719de1/artifacts. Currently registered schemes are: ['', 'file', 's3', 'gs', 'wasbs', 'ftp', 'sftp', 'dbfs', 'hdfs', 'viewfs', 'runs', 'models']

Could you please tell me how can I reproduce the Scenario 5 where the client code cannot access Minio, but actually uploads the artifacts on Minio passing them to the Mlflow server?

Thank you in advance for any help you can give me.

**Comment by rgaudenzi-fub** [comment]
Please notice that the MinIO credentials passed to the MLflow server are correct: in fact, if I try to access an already existing file (e.g., uploaded manually on Minio outside MLflow) using the REST API (e.g., localhost/api/2.0/mlflow-artifacts/artifacts/numbers.txt) then I get the file successfully.

The problem is that the client code just can't upload its artifacts without accessing the bucket directly (using the scheme described in Scenario 4 on the documentation (https://mlflow.org/docs/latest/tracking.html#scenario-4-mlflow-with-remote-tracking-server-backend-and-artifact-stores).

**Comment by BenWilson2** [comment]
Hi @rgaudenzi-fub could you make sure that you're installing the latest version of MLflow? This functionality only became available recently in version 1.24.0, released on 2/28/22 (portions of it were released in 1.23.x, but there were a few issues with it, so please use 1.24.0). 
The exception that you posted would only be generated by an MLflow version < 1.23.0.

**Comment by rgaudenzi-fub** [comment]
@BenWilson2 Hi Ben, thank you for your help. I noticed there was a new version and installed it on server-side, but unfortunately I didn't think to update MLflow on client-side. Now it works without any problem. Thanks!

**Comment by ThorHe** [comment]
@rgaudenzi-fub which version of `--artifacts-destination` and/or `--default-artifact-root` ended up working for you?

**Comment by rgaudenzi-fub** [comment]
Hi @ThorHe, of course, I'll share anything that could help.

On server-side, I start MLflow using the following commands:

export MLFLOW_S3_ENDPOINT_URL="<your_s3_endpoint_url>"
export AWS_ACCESS_KEY_ID="<your_s3_username>"
export AWS_SECRET_ACCESS_KEY="<your_s3_password>"
mlflow server \
--backend-store-uri mysql+mysqlconnector://$DB_USER:$DB_PWD@db:3306/mlflowtest \
--artifacts-destination s3://mlflow \
--serve-artifacts \
--host 0.0.0.0 \
--gunicorn-opts "--timeout 300 --log-level debug"

Where $DB_USER and $DB_PWD are env variables containing your DB credentials.

On client-side, I did like this:
mlflow.set_tracking_uri('<your_mlflow_endpoint_base_url>')
experiment_name = "ld_multinominalnb_%s" % datetime.now()
mlflow.create_experiment(experiment_name)
mlflow.set_experiment(experiment_name)

# Here I omit some code that loads a dataset from a file and split it in training and testing sets, not related to MLFlow.

with mlflow.start_run():# Instantiating model with model parameters
    model = MultinomialNB()
    model.fit(x_train, y_train)
    y_pred = model.predict(x_test)
    ac = accuracy_score(y_test, y_pred)
    mlflow.log_metric('accuracy', ac)
    # Logging model to MLFlow
    mlflow.sklearn.log_model(sk_model = model,
                             artifact_path = 'ld-multinomialdb-model',
                             registered_model_name = 'ld-multinomialdb-model')

Please notice that MLflow needs to be updated to version 1.24.0 (or above) on both client and server side, or you'll get the same error that I had.

Hope this helped, let me know if you need more information.

**Comment by ThorHe** [comment]
Thank you for this writeup @rgaudenzi-fub! It really helped in getting the setup right.

I had to make a few changes, as I'm using [Azure Blob Storage](https://www.mlflow.org/docs/latest/tracking.html#azure-blob-storage) to host my artifacts. 
1. I have not supplied any other storage related variables other than `AZURE_STORAGE_ACCESS_KEY` 
2. The `--artifact-destination` argument supplied with the following syntax:
`wasbs://<container name>@<storage account name>.blob.core.windows.net/<mlflow root path>`. 

The docs are a bit unclear whether the `s3://mlflow` syntax could also be applied to other storage implementations.

What is your reasoning for creating a new experiment for each run, instead of just doing multiple runs in the same experiment?

**Comment by rgaudenzi-fub** [comment]
@ThorHe You're welcome, I'm glad to share what I've done with others who faced the same problems :)

About the experiment, there is no real reasoning to be honest: the code I pasted here was just a proof-of-concept that I did to show MLflow to my colleagues at work and test the deploy on a remote server. 

If you've found any guide o tutorial that shows how to use MLflow properly on these matters I'd love to read it, it could help me a lot.

**Comment by cgebe** [comment]
I am facing a similar issue getting the following error:

```
(InvalidAccessKeyId) when calling the PutObject operation: The AWS Access Key Id you provided does not exist in our records.
```

My entrypoint for the server:

```
ENTRYPOINT mlflow server \
    --host 0.0.0.0 \
    --backend-store-uri postgresql://$DB_USER:$DB_PASSWORD@$DB_HOST/$DB_NAME \
    --artifacts-destination $ARTIFACT_DESTINATION --serve-artifacts \
    --default-artifact-root $ARTIFACT_ROOT \ 
    --gunicorn-opts "--log-level debug" 
```

My environment vars:

```
          - name: ARTIFACT_ROOT
            value: "https://mlflow.foo.bar.com/api/2.0/mlflow-artifacts/artifacts/"
          - name: ARTIFACT_DESTINATION
            value: "s3://s3.eu-central-003.backblazeb2.com/mlflowbucket"
          - name: AWS_ACCESS_KEY_ID
            valueFrom:
              secretKeyRef:
                name: mlflow-secrets
                key: s3-key
          - name: AWS_SECRET_ACCESS_KEY
            valueFrom:
              secretKeyRef:
                name: mlflow-secrets
                key: s3-secret
```

What am i supposed to set here? When i set `MLFLOW_S3_ENDPOINT_URL` to `"https://s3.eu-central-003.backblazeb2.com"` the client basically appends the `ARTIFACT_DESTINATION` to the `MLFLOW_S3_ENDPOINT_URL` which naturally results in an error. The documentation is unclear at this point and basically says to set something here and there... Could you please give more information, what entire set of env vars worked in your case? What should be set and what omitted? Would appreciate that!

**Comment by cgebe** [comment]
To answer my own question:

Make sure to set the `ARTIFACT_DESTINATION` to just the scheme and bucket without netloc -> `s3://bucket/path`. In my case setting `"s3://s3.eu-central-003.backblazeb2.com/mlflowbucket"` to `s3://mlflowbucket` fixed it for me, after debugging the code and seeing how the boto3 client is actually called...

**Comment by davidhyun** [comment]
@rgaudenzi-fub 
It works for me. thank you.
Almost examples on web save artifact at local direcotry. This is not my case.
I want to know that scenario 5 on server-side and client-side specific settings.
Thank you. Thank you.
